pkg/util/machines.go

```diff
@@ -114,12 +114,24 @@ func IsControlPlaneMachine(machine metav1.Object) bool {
 // GetMachineMetadata returns the cloud-init metadata as a base-64 encoded
 // string for a given VSphereMachine.
 func GetMachineMetadata(hostname string, vsphereVM infrav1.VSphereVM, networkStatuses ...infrav1.NetworkStatus) ([]byte, error) {
+	return GetMachineMetadataWithIPAMState(hostname, vsphereVM, map[string]infrav1.NetworkDeviceSpec{}, networkStatuses...)
+}
+
+// GetMachineMetadataWithIPAMState the cloud-init metadata as a base-64 encoded
+// string for a given VSphereMachine.
+// IPAM state includes IP and Gateways that should be added to each device.
+func GetMachineMetadataWithIPAMState(hostname string, vsphereVM infrav1.VSphereVM, ipamState map[string]infrav1.NetworkDeviceSpec, networkStatuses ...infrav1.NetworkStatus) ([]byte, error) {
 	// Create a copy of the devices and add their MAC addresses from a network status.
 	devices := make([]infrav1.NetworkDeviceSpec, integer.IntMax(len(vsphereVM.Spec.Network.Devices), len(networkStatuses)))
 
 	var waitForIPv4, waitForIPv6 bool
 	for i := range vsphereVM.Spec.Network.Devices {
 		vsphereVM.Spec.Network.Devices[i].DeepCopyInto(&devices[i])
+		if state, ok := ipamState[devices[i].MACAddr]; ok {
+			devices[i].IPAddrs = append(devices[i].IPAddrs, state.IPAddrs...)
```

Comment:
this might not work with clusterClass, this is a case of co-authored slices. we need to be able a unique key so that server-side apply works. cc @fabriziopandini

Reply:
Confirmed.
if order to allow one contreller (api) to create the device entry, and another (capv) to set the ipAddr into device entries, the list of devices must be defined as +ListMap,

Reply:
ok, just confirmed it seems like we're never retrofitting `.spec.devices.ipAddrs` so it shouldn't be an issue. 

@tylerschultz  can you confirm we're not patching `.spec.devices.ipAddrs` of the vsphereVM with the result of the fulfilled claim?

Reply:
We had originally updated `spec.devices.ipAddrs` on the VSphereVM, but we realized we were modifying the spec of the object so we backed it out. We changed it to use the in-memory map instead. We do not modify the `.spec.devices.ipAddrs`.